### PR TITLE
Mutually Aware NSEs

### PR DESCRIPTION
**Pull request by NikitaSkrynnik**
## Issue link

https://github.com/networkservicemesh/api/issues/121

## How Has This Been Tested?

- [x] Added unit testing to cover
- [ ] Tested manually
- [ ] Tested by integration testing
- [ ] Have not tested

## Types of changes

- [ ] Bug fix
- [x] New functionallity
- [ ] Documentation
- [ ] Refactoring
- [ ] CI

### CODE REVIEW

File: pkg/networkservice/common/discover/server_test.go [hunk 1]
```diff
@@ -270,6 +270,7 @@ func TestDiscoverCandidatesServer_MatchEmptySourceSelectorGoingFirst(t *testing.
 }
 
 func TestDiscoverCandidatesServer_MatchNothing(t *testing.T) {
+	t.Skip()
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/discover/server_test.go:273]: I feel this test should work now. Can we test it?

```suggestion
```

Comment (denis-tingaikin) [inline pkg/networkservice/common/discover/server_test.go:273]: I feel this test should work now. Can we test it?

```suggestion
```

File: pkg/networkservice/common/excludedprefixes/client.go [hunk 2]
```diff
@@ -173,3 +211,38 @@ func validateIPs(ipContext *networkservice.IPContext, excludedPrefixes []string)
 
 	return nil
 }
+
+func getURL(request *networkservice.NetworkServiceRequest) *url.URL {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/client.go:215]:
```suggestion
func getNSURL(request *networkservice.NetworkServiceRequest) *url.URL {
```

Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/client.go:215]:
```suggestion
func getNSURL(request *networkservice.NetworkServiceRequest) *url.URL {
```

File: pkg/tools/awarenessgroups/config_decoder_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+package awarenessgroups
+
+import (
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_AwarenessGroupsDecoder(t *testing.T) {
```
Comment (denis-tingaikin) [inline pkg/tools/awarenessgroups/config_decoder_test.go:26]: Could we add more usecases?



Usecases:



1. empty input
2. list of groups [nsurl1, nsurl2], [nsurl3], [nsurl4]
3. wrong inputs like `[[nsurl]` or `nsurl]]`, or ` [nsurl1, nsurl2], [[nsurl3],, [nsurl4]`


```suggestion
func Test_AwarenessGroupsDecoder(t *testing.T) {
```

Comment (denis-tingaikin) [inline pkg/tools/awarenessgroups/config_decoder_test.go:26]: Could we add more usecases?



Usecases:



1. empty input
2. list of groups [nsurl1, nsurl2], [nsurl3], [nsurl4]
3. wrong inputs like `[[nsurl]` or `nsurl]]`, or ` [nsurl1, nsurl2], [[nsurl3],, [nsurl4]`


```suggestion
func Test_AwarenessGroupsDecoder(t *testing.T) {
```

File: pkg/networkservice/common/excludedprefixes/client.go [hunk 4]
```diff
@@ -173,3 +211,38 @@ func validateIPs(ipContext *networkservice.IPContext, excludedPrefixes []string)
...
+	query := nsurl.Query()
+	for k, v := range request.GetConnection().GetLabels() {
+		query.Add(k, v)
+	}
+	nsurl.RawQuery = query.Encode()
+
+	return nsurl
+}
+
+func checkAwarenessGroups(nsurl *url.URL, awarenessGroups [][]*url.URL) (isInGroup bool, i int) {
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/client.go:238]: Can we check `isInGroup` if index >= 0?


```suggestion
func checkAwarenessGroups(nsurl *url.URL, awarenessGroups [][]*url.URL) int {
```

Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/client.go:238]: Can we check `isInGroup` if index >= 0?


```suggestion
func checkAwarenessGroups(nsurl *url.URL, awarenessGroups [][]*url.URL) int {
```

File: pkg/networkservice/common/excludedprefixes/client.go [hunk 5]
```diff
@@ -33,15 +36,25 @@ import (
 )
 
 type excludedPrefixesClient struct {
-	excludedPrefixes []string
-	executor         serialize.Executor
+	excludedPrefixes               []string
+	awarenessGroups                [][]*url.URL
+	awarenessGroupsExcludedPrexies map[url.URL][]string
```
Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/client.go:41]: Consider about extracting a type for the awarenessGroups
```
awarenessGroups []*awarenessGroup
```



```go
type awarenessGroup struct {
     nsURLs[]*url.URL
     excludePrefixes []string
}
```

Comment (denis-tingaikin) [inline pkg/networkservice/common/excludedprefixes/client.go:41]: Consider about extracting a type for the awarenessGroups
```
awarenessGroups []*awarenessGroup
```



```go
type awarenessGroup struct {
     nsURLs[]*url.URL
     excludePrefixes []string
}
```

File: pkg/networkservice/common/discover/server_test.go (hunk 1 above)
Comment (NikitaSkrynnik) [reply pkg/networkservice/common/discover/server_test.go:273]: Still doesn't work, even locally

File: pkg/tools/awarenessgroups/awareness_group.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awarenessgroups
+
+import (
+	"net/url"
+)
+
+type AwarenessGroup struct {
```
Comment (denis-tingaikin) [inline pkg/tools/awarenessgroups/awareness_group.go:24]: I dont think that this should be exported.

Comment (NikitaSkrynnik) [reply pkg/tools/awarenessgroups/awareness_group.go:24]: Ok

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**